File: app/src/main/java/milespeele/canvas/drawing/CanvasPoint.java
package milespeele.canvas.drawing;

/**
 * Created by mbpeele on 11/29/15.
 */
class CanvasPoint {

    public float x, y;
    public long time;

    public CanvasPoint(float x, float y, long time) {
        this.x = x;
        this.y = y;
        this.time = time;
    }

    public float computeDistance(CanvasPoint p) {
        float dx = x - p.x;
        float dy = y - p.y;
        return (float) Math.sqrt(dx * dx + dy * dy);
    }

    public float computeVelocity(CanvasPoint p) {
        long duration = Math.abs(time - p.time);
        return (duration != 0) ? computeDistance(p) / duration : computeDistance(p);
    }

    public CanvasPoint computeMidpoint(CanvasPoint p2) {
        return new CanvasPoint((x + p2.x) / 2.0f, (y + p2.y) / 2.0f, (time + p2.time) / 2);
    }
}

File: app/src/main/java/miles/scribble/home/drawing/CanvasPoint.kt
package miles.scribble.home.drawing

import android.os.SystemClock
import miles.scribble.util.extensions.DateExtensions

/**
 * Created by mbpeele on 11/29/15.
 */
data class CanvasPoint(var x: Float, var y: Float, var time: Long = DateExtensions.currentTimeInMillis) {

    fun computeDistance(p: CanvasPoint): Float {
        val dx = x - p.x
        val dy = y - p.y
        return Math.sqrt((dx * dx + dy * dy).toDouble()).toFloat()
    }

    fun computeVelocity(p: CanvasPoint): Float {
        val duration = Math.abs(time - p.time)
        return if (duration > 0L) computeDistance(p) / duration else computeDistance(p)
    }

    fun computeMidpoint(p2: CanvasPoint): CanvasPoint {
        return CanvasPoint((x + p2.x) / 2.0f, (y + p2.y) / 2.0f, (time + p2.time) / 2)
    }
}
